src/traces/image/hover.js

```diff
@@ -0,0 +1,79 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Fx = require('../../components/fx');
+var Lib = require('../../lib');
+var constants = require('./constants');
+
+// var Axes = require('../../plots/cartesian/axes');
+
+module.exports = function hoverPoints(pointData, xval, yval) {
+    var cd0 = pointData.cd[0];
+    var trace = cd0.trace;
+    var xa = pointData.xa;
+    var ya = pointData.ya;
+
+    // Return early if not on image
+    if(Fx.inbox(xval - cd0.x0, xval - (cd0.x0 + cd0.w * trace.dx), 0) > 0 ||
+            Fx.inbox(yval - cd0.y0, yval - (cd0.y0 + cd0.h * trace.dy), 0) > 0) {
+        return;
+    }
+
+    // Find nearest pixel's index and pixel center
+    var nx = Math.floor((xval - cd0.x0) / trace.dx);
+    var ny = Math.floor(Math.abs(yval - cd0.y0) / trace.dy);
+
+    var hoverinfo = cd0.hi || trace.hoverinfo;
+    var fmtColor;
+    if(hoverinfo) {
+        var parts = hoverinfo.split('+');
+        if(parts.indexOf('all') !== -1) parts = ['color'];
+        if(parts.indexOf('color') !== -1) fmtColor = true;
+    }
+
+    var colormodel = trace.colormodel;
+    var dims = colormodel.length;
+    var c = trace._scaler(cd0.z[ny][nx]);
+    var s = constants.colormodel[colormodel].suffix;
+
+    var colorstring = [];
+    if(trace.hovertemplate || fmtColor) {
+        colorstring.push('[' + [c[0] + s[0], c[1] + s[1], c[2] + s[2]].join(', '));
+        if(dims === 4) colorstring.push(', ' + c[3] + s[3]);
+        colorstring.push(']');
+        colorstring = colorstring.join('');
+        pointData.extraText = '<span style="text-transform:uppercase">' + colormodel + '</span>: ' + colorstring;
+    }
+
+    var py = ya.c2p(cd0.y0 + (ny + 0.5) * trace.dy);
+    var xVal = cd0.x0 + (nx + 0.5) * trace.dx;
+    var yVal = cd0.y0 + (ny + 0.5) * trace.dy;
```

Comment:
Does `image` work on categorical axes? If so, I suspect this part here will break.

Reply:
It doesn't at the moment. 

However, I added support to it in this [throwaway branch](https://github.com/plotly/plotly.js/compare/pr-image...antrg-test) along with improvements for log axis type. You can see it in action in this [Codepen](https://codepen.io/antoinerg/pen/BaaWjxe).

It shows that:
- `x0` and `y0` are working
- now if `dx` or `dy` is not equal to 1, we may get strange results similar to what `heatmap` currently does.

@etpinard is this good enough?

Reply:
The behaviour looks good @antoinerg -- thanks!

One tip: use `d2c` in `calc.js`, for example:

https://github.com/plotly/plotly.js/blob/82e33e2491192eacab1db026af7b2b98cc6e14ab/src/traces/image/calc.js#L17-L20

should just be:

```js
x0 = xa.d2c(trace.x0) - trace.dx / 2;
y0 = ya.d2c(trace.y0) - trace.dy / 2;
```

where `d2c` for linear and log axes is simply an is-numeric check.

Reply:
:ok_hand: done in fa04037cdad8a130068f08f0af417915b0147015